### ISSUE TITLE
Problem with total  observed statistics in mixture model

### ISSUE THREAD

**Issue by RMKirkpatrick** [OpenMx/OpenMx#398]
I discovered yesterday that one of my demonstration scripts,
[mixture_Pboot.R.txt](https://github.com/user-attachments/files/17777928/mixture_Pboot.R.txt)
, from the AGES 2017 Workshop is broken with the current OpenMx release.  In this script, the two mixture components in the two-class mixture model are MxRAMModels, neither of which contain an MxData object.  What appears to be happening in the backend is that each MxRAMModel is seeing the dataset of 10,000 observed statistics in the "container" model, and "remembering" the 10,000 observed statistics.  Then, in the frontend, when MxSummary counts the total number of observed statistics, the 10,000 is being counted twice, because both MxRAMModels report it.  This throws off the degrees-of-freedom, which in turn throws off the use of `mxCompare()` on the 2- and 1-component mixture models...and `mxCompare()` is the UI for the bootstrap likelihood-ratio test.

**Comment by mcneale** [comment]
Odd that, I thought that we had addressed the duplicate dataset number of statistics issue.  In part, I thought it was because the statistics were only counted if they were not (automatically) inherited from a parent.

Best,
Mike

Michael C. Neale, Ph.D.
Banks Distinguished Professor of Psychiatry
Virginia Institute for Psychiatric and Behavioral Genetics
Virginia Commonwealth University
Email: ***@***.***
Zoom: https://zoom.us/my/michael.neale

On Nov 15, 2024, at 11:04 AM, RMKirkpatrick ***@***.***> wrote:


This Message Is From an External Sender
This message came from outside your organization.
<https://us-phishalarm-ewt.proofpoint.com/EWT/v1/JqxBPMk!llOjrVxo-zBbsVuX23FMbBDcRxkGuo-Kmsb37Z8YEGpbQ3VZJL07ieh3qSTeT1v1LMoxTSCddH8nS6MT_fw2$>
Report Suspicious


I discovered yesterday that one of my demonstration scripts,
mixture_Pboot.R.txt<https://urldefense.com/v3/__https://github.com/user-attachments/files/17777928/mixture_Pboot.R.txt__;!!JqxBPMk!k0kdD_rcLVj1vAq5YWPKvyZtEAdPBWZZKWlNH86tFUVsSWSKIgV7nI4jgKUwcsbdGox5NK0geC_adQtpT_88-qiLF0M$>
, from the AGES 2017 Workshop is broken with the current OpenMx release. In this script, the two mixture components in the two-class mixture model are MxRAMModels, neither of which contain an MxData object. What appears to be happening in the backend is that each MxRAMModel is seeing the dataset of 10,000 observed statistics in the "container" model, and "remembering" the 10,000 observed statistics. Then, in the frontend, when MxSummary counts the total number of observed statistics, the 10,000 is being counted twice, because both MxRAMModels report it. This throws off the degrees-of-freedom, which in turn throws off the use of mxCompare()on the 2- and 1-component mixture models...and mxCompare() is the UI for the bootstrap likelihood-ratio test.

Note that models/passing/SummaryCheck.R still passes, because the mixture components in that script are not RAM.

**Comment by RMKirkpatrick** [comment]
The issue applies even if the mixture components use MxExpectationNorrmal instead of RAM expectation.

**Comment by RMKirkpatrick** [comment]
I used `git bisect` today, and my suspicion turned out to be true: 24072d3 is the source of this issue.

**Comment by RMKirkpatrick** [comment]
I discovered today that the cause of this issue is occurring during the MxComputeReportExpectation step of the compute plan.  I also discovered that modifying `MVNExpectation::numObservedStats()` to unconditionally return `0` does not resolve the issue.

Then, at the current head of master (cf7f7da), I made this change:
```
diff --git a/src/omxExpectation.cpp b/src/omxExpectation.cpp
index 5cb9dedb..d335dfb6 100644
--- a/src/omxExpectation.cpp
+++ b/src/omxExpectation.cpp
@@ -644,6 +644,6 @@ int MVNExpectation::numObservedStats()
 
 void MVNExpectation::populateAttr(SEXP algebra)
 {
-  IntegerVector RobStat = Rcpp::wrap(numObservedStats());
-  Rf_setAttrib(algebra, Rf_install("numStats"), RobStat);
+  //IntegerVector RobStat = Rcpp::wrap(numObservedStats());
+  //Rf_setAttrib(algebra, Rf_install("numStats"), RobStat);
 }
```
That change stops OpenMx from redundantly counting observed statistics.  Somewhat surprisingly, that change also does not break any tests in 'make test', which has the same results with or without the change:
```
Number of errors: 4 
Error SLSQP models/passing/StateSpaceErrorCheck.R ***
[1] "MxExpectationStateSpace: row frequencies or weights provided in 'State Space Error Check.data' are not supported"

Error CSOLNP models/passing/StateSpaceErrorCheck.R ***
[1] "MxExpectationStateSpace: row frequencies or weights provided in 'State Space Error Check.data' are not supported"

Error CSOLNP models/passing/UselessConstraint.R ***
[1] "`a` is not TRUE\n\n`actual`:   FALSE\n`expected`: TRUE "

Error NPSOL models/passing/StateSpaceErrorCheck.R ***
[1] "MxExpectationStateSpace: row frequencies or weights provided in 'State Space Error Check.data' are not supported"

Finished testing models.
make: *** [Makefile:193: test] Error 4
```
The failure in models/passing/StateSpaceErrorCheck.R is awaiting a patch from @mhunter1 .  The optimizer-specific failure in models/passing/UselessConstraint.R is new, and based on a comment above the failing test, I am not sure it is benign.

Anyhow, as far as this issue is concerned, it appears that asking MVNExpectations (of which MxExpectationRAM is a type) to count their observed statistics during the  MxComputeReportExpectation step causes data to be counted redundantly.

**Comment by RMKirkpatrick** [comment]
I just tried 'make nightly' with the above change to src/omxExpectation.cpp .  The job did not finish due to a segfault from models/nightly/StateSpaceLatentGrowth.R , which also segfaults without the change.  The only test failure I observed that was due to the change was `omxCheckEquals(bivSatSummary4$observedStatistics, 5)` in demo/BivariateSaturated_MatrixRaw.R ; the change causes OpenMx to report 2000, not 5, observed statistics for the model, and 2000 is clearly the correct value for a raw dataset with 1000 rows and 2 columns.

**Comment by RMKirkpatrick** [comment]
`git bisect` indicates that the aforementioned segfault results from a change made to models/nightly/StateSpaceLatentGrowth.R in de60f13.  It looks as though this fault may be platform-specific.  @mhunter1 ?

**Comment by mhunter1** [comment]
@RMKirkpatrick That commit is adding a test that demonstrates a problem, not a change that creates a problem.  See #389

**Comment by RMKirkpatrick** [comment]
> @RMKirkpatrick That commit is adding a test that demonstrates a problem, not a change that creates a problem. See https://github.com/OpenMx/OpenMx/issues/389

Aha!  Now I see.

**Comment by tbates** [comment]
This will address a comment from people using other software surprised to see the row num instead of varnum as the stat. But I seem also to recall (like a decade ago) some here arguing that the larger number was the right stat to count?

**Comment by RMKirkpatrick** [comment]
I resolved #389 earlier this afternoon.  It was important to take care of it, because the segfault in question crashes R and thereby kills a 'make nightly' job then and there.

**Comment by mcneale** [comment]
Good work Rob!

**Comment by RMKirkpatrick** [comment]
Resolved with 93b2a48.  Failing test corrected with 0d22ed6.  Regression tests added with 3b22ca7.
